### PR TITLE
ROX-12844: Refactor sensor image enrichment with multiple image names

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

[The parent PR has the initial problem statement as well as the references to all follow-up PRs that have been done, please consult it to get the full context of changes being made here.](https://github.com/stackrox/stackrox/pull/3511)

Additionally, [this doc exists for a overview of the changes being made within this series of PRs.](https://docs.google.com/document/d/1xxBQMRGQHOJg1E44iCUkMOGul7ht_496VubhSZYJ42k/edit?usp=sharing)

Within this PR, the sensor will now re-scan images that have the same digest, but have different image names associated with them. This will lead to central correctly upserting and re-fetching signatures & signature verification data for the newly added image name. It will also ensure that this will be continuously reprocessed.

Note that there will be more follow-ups coming, specifically:
- Changing the image signature policy to take the image full name of the deployment into account when verifying signatures.
- Adding E2E tests that cover the cases with images with the same digest but different image names.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

Manual tests (note, that automated E2E tests will be added in a follow-up, essentially emulating whatever I have done here now + the policy part):
```bash
# The test setup requires two images, located in two different repositories within a container registry, e.g. dockerhub.
# I have prepped two images for the manual tests specifically, with [crane](https://github.com/google/go-containerregistry/blob/main/cmd/crane/doc/crane.md) we can check that they have the same digest:
crane digest  daha97/nginx:1.23
sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385

crane digest daha97/alt-nginx:1.23
sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385

# (As an additional note, only `alt-nginx` has a signature associated with it).

# Start with creating a pod running the first image:
k run --image=daha97/nginx:1.23

# Observe within the API that the image has been created successfully + enriched:
roxcurl /v1/images | jq
....
    {
      "id": "sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385",
      "name": "docker.io/daha97/nginx:1.23",
      "components": 108,
      "cves": 72,
      "fixableCves": 0,
      "created": "2022-11-15T13:14:32.170016733Z",
      "lastUpdated": "2022-11-25T03:11:11.222162138Z",
      "priority": "1"
    }
....

# Inspecting the image closer should reveal that no signature data is associated with it:
roxcurl /v1/images/sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385 | jq -r '.notes'
  "notes": [
    "MISSING_SIGNATURE",
    "MISSING_SIGNATURE_VERIFICATION_DATA"
  ]

# We also observe the "names" field being populated correctly, but only with a single name:
roxcurl /v1/images/sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385 | jq -r '.names'
[
  {
    "registry": "docker.io",
    "remote": "daha97/nginx",
    "tag": "1.23",
    "fullName": "docker.io/daha97/nginx:1.23"
  }
]

# Now, create another pod running the second image:
k run --image=daha97/alt-nginx:1.23

# Observe within the names field that this now contains two entries:
roxcurl /v1/images/sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385 | jq -r '.names'
[
  {
    "registry": "docker.io",
    "remote": "daha97/nginx",
    "tag": "1.23",
    "fullName": "docker.io/daha97/nginx:1.23"
  },
  {
    "registry": "docker.io",
    "remote": "daha97/alt-nginx",
    "tag": "1.23",
    "fullName": "docker.io/daha97/alt-nginx:1.23"
  }
]

# Additionally, we see that the image signature will be fetched. If we now configure a signature integration, this would also be verified:
roxcurl /v1/images/sha256:6ad8394ad31b269b563566998fd80a8f259e8decf16e807f8310ecc10c687385 | jq -r '.notes'
  "notes": [
    "MISSING_SIGNATURE_VERIFICATION_DATA"
  ]
```

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3907** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3907" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/3907?utm_source=stack-comment).

File: central/image/service/service_impl.go
Comment (ivan-degtiarenko) [inline central/image/service/service_impl.go:281]: Are we sure saving the result of this `enricher.EnrichImage` when:
1. Error is returned
2. Image is already in the datastore but request image name isn't https://github.com/stackrox/stackrox/pull/3907/files#diff-1969ddcb957be51b7db622434fa6b73050a9b853fe85c781790b74c7fc55078cR235

is the correct approach?

Before this PR, this code path only ignored image when there didn't exist an image already in datastore. Now we can actually replace the existing image.

File: sensor/common/detector/enricher.go
Comment (ivan-degtiarenko) [inline sensor/common/detector/enricher.go:158]:
```suggestion
	// In case of image full name, we can skip. In case of image ID, we should make sure to check if the image's name
```

File: central/image/service/service_impl.go
Comment (dhaus67) [reply central/image/service/service_impl.go:281]: I think that's a fair point and good catch.

I've re-written `ScanImageInternal` to skip upserting the image in the database when:
- the image exists previously.
- the error from enriching the image is `!= nil`.

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM modulo fixing style checks

File: central/image/service/service_impl.go
Comment (ivan-degtiarenko) [inline central/image/service/service_impl.go:228]: I think this `if` is not covered by https://github.com/stackrox/stackrox/pull/3950, right? Maybe it makes sense to introduce a test here

**Comment by dhaus67** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/3907?utm_source=stack-comment).


Job Reference: Lqz66Yqufrzx2FHkfF78

**Pull request merged**
